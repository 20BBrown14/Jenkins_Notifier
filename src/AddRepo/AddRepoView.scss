.AddRepoview-Window {
  width: 650px;
  height: 450px;
}