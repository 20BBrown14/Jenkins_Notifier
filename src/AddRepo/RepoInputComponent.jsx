import React from 'react';
import InputField from 'terra-form-input/lib/InputField';
import PropTypes from 'prop-types';


const propTypes = {
/** Error message */
  errorMessage: PropTypes.string.isRequired,
  /** Whether form input is invalid */
  formInvalid: PropTypes.bool,
  /** Function for when repo url input field changes */
  URLFieldChange: PropTypes.func.isRequired,
  /** Function for when repo name input field changes */
  nameFieldChanged: PropTypes.func.isRequired,
  /** Help message to display under input field */
  helpMessage: PropTypes.string,
  /** Value of the repo URL input field */
  repoURL: PropTypes.string.isRequired,
  /** Value of the repo name input field */
  repoName: PropTypes.string.isRequired,
  /** Whether or not the view is loading */
  isLoading: PropTypes.bool,
};

const defaultProps = {
  formInvalid: false,
};

const RepoInputComponent = (props) => {
  const {
    errorMessage,
    formInvalid,
    URLFieldChange,
    helpMessage,
    repoURL,
    repoName,
    nameFieldChanged,
    isLoading,
  } = props;
  return (
    <div>
      <InputField
        style={{ margin: '15px 25px 12.5px 25px' }}
        inputId="repoNameInput"
        label="Name"
        value={repoName}
        onChange={nameFieldChanged}
        help="Input no name to use name on Jenkins"
        inputAttrs={{ disabled: isLoading }}
      />
      <InputField
        style={{ margin: '12.5px 25px 12.5px 25px' }}
        inputId="repoURLInput"
        label="URL"
        error={errorMessage}
        help={helpMessage}
        isInvalid={formInvalid}
        onChange={URLFieldChange}
        value={repoURL}
        inputAttrs={{ disabled: isLoading }}
      />
    </div>
  );
};

RepoInputComponent.propTypes = propTypes;
RepoInputComponent.defaultProps = defaultProps;

export default RepoInputComponent;
