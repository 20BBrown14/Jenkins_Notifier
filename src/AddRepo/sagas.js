import { put, call, takeLatest, take, race } from 'redux-saga/effects';
import { A_VALIDATE_REPO_URL, invalidRepoURL, A_VALID_REPO_URL, A_CANCEL_CLICKED } from './actions';
import request from '../modules/dataLoadSagas';

/**
 * Saga to validate a repo url
 * @param {object} action
 */
export function* validateRepo(action) {
  try {
    const repoURL = action.data.URL;
    if (!repoURL.endsWith('/api/json') && !repoURL.endsWith('/api/json/')) {
      yield put(invalidRepoURL('URL should be appended with \'/api/json\''));
      return;
    }
    const { response } = yield race({
      response: call(request, action.data.URL),
      cancel: take(A_CANCEL_CLICKED),
    });
    if (response) {
      if (response.data) {
        yield put({ type: A_VALID_REPO_URL, data: { jsonData: response.data } });
      } else {
        yield put(invalidRepoURL(response.err.message));
      }
    }
  } catch (e) {
    yield put(invalidRepoURL('Error making network request'));
  }
}

function* PopupSaga() {
  yield takeLatest(A_VALIDATE_REPO_URL, validateRepo);
}

export default [PopupSaga];
