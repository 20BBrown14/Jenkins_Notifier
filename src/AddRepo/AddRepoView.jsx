import React from 'react';
import PropTypes from 'prop-types';
import Button from 'terra-button';
import Text from 'terra-text';
import IconSuccess from 'terra-icon/lib/icon/IconSuccess';
import Image from 'terra-image';
import RepoInputComponent from './RepoInputComponent';
import './AddRepoView.scss';
import loadingSpinner from '../assets/loading_Spinner.gif';

const propTypes = {
  /** Error message for input */
  errorMessage: PropTypes.string.isRequired,
  /** If input is invalid */
  formInvalid: PropTypes.bool,
  /** Function for when repo url input field changes */
  URLFieldChange: PropTypes.func.isRequired,
  /** Function for when repo name input field changes */
  nameFieldChanged: PropTypes.func.isRequired,
  /** Help message to display under input field */
  helpMessage: PropTypes.string,
  /** Value of the repo URL input field */
  repoURL: PropTypes.string.isRequired,
  /** Value of the repo name input field */
  repoName: PropTypes.string.isRequired,
  /** Handler for when validate is clicked */
  validateClickHandler: PropTypes.func,
  /** Handler for when confirm is clicked */
  confirmClickHandler: PropTypes.func,
  /** Whether or not the URL has been validated */
  validated: PropTypes.bool,
  /** Whether or not the URL has been confirmed */
  confirmed: PropTypes.bool,
  /** Function for when the cancel button is clicked */
  cancelClickHandler: PropTypes.func.isRequired,
  /** Bool to determine if the app is loading or not */
  isLoading: PropTypes.bool.isRequired,
};

const defaultProps = {
  formInvalid: false,
};

const AddRepoView = (props) => {
  const {
    errorMessage,
    formInvalid,
    URLFieldChange,
    nameFieldChanged,
    helpMessage,
    repoURL,
    validateClickHandler,
    confirmClickHandler,
    repoName,
    validated,
    confirmed,
    cancelClickHandler,
    isLoading,
  } = props;

  return (
    <div className="AddRepoview-Window">
      <Text style={{ margin: '12px 0px 0px 25px' }} fontSize={32}>Track a New Repo</Text>
      {(isLoading) && (
        <Image
          src={loadingSpinner}
          height="60"
          width="60"
          style={{ top: '7px', position: 'absolute' }}
        />
      )}
      <RepoInputComponent
        errorMessage={errorMessage}
        formInvalid={formInvalid}
        URLFieldChange={URLFieldChange}
        nameFieldChanged={nameFieldChanged}
        helpMessage={helpMessage}
        repoURL={repoURL}
        repoName={repoName}
        isLoading={isLoading}
      />
      <Button
        style={{ margin: '0px 5px 0px 25px' }}
        text="Validate"
        onClick={validateClickHandler}
        isDisabled={validated || isLoading}
      />
      <Button
        style={{ margin: '0px 5px 0px 5px' }}
        text="Confirm"
        onClick={confirmClickHandler}
        isDisabled={confirmed || isLoading}
      />
      <Button
        style={{ margin: '0px 5px 0px 5px' }}
        text="Cancel"
        onClick={cancelClickHandler}
      />
      {confirmed && validated &&
        <IconSuccess />}
    </div>
  );
};

AddRepoView.propTypes = propTypes;
AddRepoView.defaultProps = defaultProps;

export default AddRepoView;
