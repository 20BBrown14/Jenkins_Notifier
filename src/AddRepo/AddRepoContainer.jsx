import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import AddRepoView from './AddRepoView';
import {
  URLInputDataChanged,
  validateRepoURL,
  nameInputDataChanged,
  A_CONFIRM_BUTTON_CLICKED,
  A_INFORMATION_CONFIRMED,
  A_CANCEL_CLICKED,
} from './actions';
import { POPUP_STATE } from './reducers';

const propTypes = {
  /** If form input is invalid */
  formInvalid: PropTypes.bool,
  /** Function for when repo url input field changes */
  URLFieldChange: PropTypes.func.isRequired,
  /** Function for when repo name input field changes */
  nameFieldChanged: PropTypes.func.isRequired,
  /** Value of the repo URL input field */
  repoURL: PropTypes.string,
  /** Value of the repo name input field */
  repoName: PropTypes.string,
  /** Handler for when validate is clicked */
  validateClickHandler: PropTypes.func,
  /** handler for when confirm is clicked */
  confirmClickHandler: PropTypes.func,
  /** Help message */
  helpMessage: PropTypes.string.isRequired,
  /** Error message to display on error */
  errorMessage: PropTypes.string,
  /** Whether or not the url has been validated */
  validated: PropTypes.bool.isRequired,
  /** Whether or not the confirmed button has been clicked */
  confirmed: PropTypes.bool.isRequired,
  /** Function for when repo information has been confirmed */
  informationConfirmed: PropTypes.func,
  /** Function for when the cancel button is clicked */
  cancelClickHandler: PropTypes.func.isRequired,
  /** Bool for if the app is loading or not */
  isLoading: PropTypes.bool.isRequired,
};

/* eslint-disable-next-line react/prefer-stateless-function */
export class AddRepoContainer extends React.Component {
  componentDidUpdate() {
    const {
      validated,
      confirmed,
      informationConfirmed,
      repoName,
    } = this.props;
    if (validated && confirmed) {
      informationConfirmed(repoName);
    }
  }

  render() {
    const {
      formInvalid,
      URLFieldChange,
      nameFieldChanged,
      repoURL,
      validateClickHandler,
      confirmClickHandler,
      helpMessage,
      repoName,
      errorMessage,
      validated,
      confirmed,
      cancelClickHandler,
      isLoading,
    } = this.props;

    const validateClick = () => {
      validateClickHandler(repoURL);
    };

    const confirmClick = () => {
      confirmClickHandler(validated, repoURL);
    };

    return (
      <AddRepoView
        errorMessage={errorMessage}
        formInvalid={formInvalid || false}
        URLFieldChange={URLFieldChange}
        nameFieldChanged={nameFieldChanged}
        helpMessage={helpMessage}
        repoURL={repoURL}
        repoName={repoName}
        validateClickHandler={validateClick}
        confirmClickHandler={confirmClick}
        validated={validated}
        confirmed={confirmed && validated}
        cancelClickHandler={cancelClickHandler}
        isLoading={isLoading}
      />
    );
  }
}

const mapStateToProps = state => ({
  repoURL: state[POPUP_STATE].repoURL,
  repoName: state[POPUP_STATE].repoName,
  helpMessage: state[POPUP_STATE].helpMessage,
  formInvalid: state[POPUP_STATE].formInvalid,
  errorMessage: state[POPUP_STATE].errorMessage,
  validated: state[POPUP_STATE].validated,
  confirmed: state[POPUP_STATE].confirmed,
  isLoading: state[POPUP_STATE].isLoading,
});

export const mapDispatchToProps = dispatch => ({
  URLFieldChange: (newValue) => {
    dispatch(URLInputDataChanged(newValue.target.value));
  },
  nameFieldChanged: (newValue) => {
    dispatch(nameInputDataChanged(newValue.target.value));
  },
  validateClickHandler: (repoURL) => {
    dispatch(validateRepoURL(repoURL));
  },
  confirmClickHandler: (validated, repoURL) => {
    dispatch({ type: A_CONFIRM_BUTTON_CLICKED });
    if (!validated) {
      dispatch(validateRepoURL(repoURL));
    }
  },
  informationConfirmed: (repoName) => {
    dispatch({ type: A_INFORMATION_CONFIRMED, data: { repoName } });
    dispatch({ type: A_CANCEL_CLICKED });
  },
  cancelClickHandler: () => {
    dispatch({ type: A_CANCEL_CLICKED });
  },
});

AddRepoContainer.propTypes = propTypes;

export default connect(mapStateToProps, mapDispatchToProps)(AddRepoContainer);
