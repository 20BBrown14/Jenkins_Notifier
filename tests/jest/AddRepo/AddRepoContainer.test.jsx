import React from 'react';
import { AddRepoContainer, mapDispatchToProps } from '../../../src/AddRepo/AddRepoContainer';
import {
  A_VALIDATE_REPO_URL,
  A_URL_INPUT_DATA_CHANGED,
  A_NAME_INPUT_DATA_CHANGED,
  A_CONFIRM_BUTTON_CLICKED,
  A_INFORMATION_CONFIRMED,
  A_CANCEL_CLICKED,
} from '../../../src/AddRepo/actions';

describe('AddRepoContainer', () => {
  describe('Initialization', () => {
    it('should render a default view', () => {
      const testContainer = (shallow(
        <AddRepoContainer
          inputValidationFailed={() => {}}
          URLFieldChange={() => {}}
          nameFieldChanged={() => {}}
          cancelClickHandler={() => {}}
          repoURL=""
          repoName=""
          helpMessage=""
          errorMessage=""
          validated={false}
          confirmed={false}
          isLoading={false}
        />,
      ));
      expect(testContainer).toMatchSnapshot();
    });
  });
  describe('Action dispatch', () => {
    let mockDispatch;
    afterEach(() => {
      expect(mockDispatch).toHaveBeenCalled();
      mockDispatch = undefined;
    });
    it('dispatches a URL input data changed action', () => {
      mockDispatch = jest.fn((action) => {
        expect(action.type).toEqual(A_URL_INPUT_DATA_CHANGED);
        expect(action.data.newValue).toEqual('newValue');
      });
      const dispatchProps = mapDispatchToProps(mockDispatch);
      const mockEventObject = { target: { value: 'newValue' } };
      dispatchProps.URLFieldChange(mockEventObject);
      expect(mockDispatch).toHaveBeenCalledTimes(1);
    });
    it('dispatches a name input data changed action', () => {
      mockDispatch = jest.fn((action) => {
        expect(action.type).toEqual(A_NAME_INPUT_DATA_CHANGED);
        expect(action.data.newValue).toEqual('newValue');
      });
      const dispatchProps = mapDispatchToProps(mockDispatch);
      const mockEventObject = { target: { value: 'newValue' } };
      dispatchProps.nameFieldChanged(mockEventObject);
      expect(mockDispatch).toHaveBeenCalledTimes(1);
    });
    it('dispatches a validate job URL action', () => {
      mockDispatch = jest.fn((action) => {
        expect(action.type).toEqual(A_VALIDATE_REPO_URL);
      });
      const dispatchProps = mapDispatchToProps(mockDispatch, {});
      dispatchProps.validateClickHandler('someURL');
      expect(mockDispatch).toHaveBeenCalledTimes(1);
    });
    describe('informationConfirmed', () => {
      it('dispatches an information confirmed action', () => {
        mockDispatch = jest.fn();
        const dispatchProps = mapDispatchToProps(mockDispatch, {});
        dispatchProps.informationConfirmed('someName');
        expect(mockDispatch).toHaveBeenCalledTimes(2);
        expect(mockDispatch.mock.calls[0][0].type).toEqual(A_INFORMATION_CONFIRMED);
        expect(mockDispatch.mock.calls[0][0].data).toEqual({ repoName: 'someName' });
      });
      it('dispatches a cancel clicked action', () => {
        mockDispatch = jest.fn();
        const dispatchProps = mapDispatchToProps(mockDispatch, {});
        dispatchProps.informationConfirmed();
        expect(mockDispatch).toHaveBeenCalledTimes(2);
        expect(mockDispatch.mock.calls[1][0].type).toEqual(A_CANCEL_CLICKED);
      });
    });
    it('dispatches a cancel clicked action', () => {
      mockDispatch = jest.fn((action) => {
        expect(action.type).toEqual(A_CANCEL_CLICKED);
      });
      const dispatchProps = mapDispatchToProps(mockDispatch, {});
      dispatchProps.cancelClickHandler();
      expect(mockDispatch).toHaveBeenCalledTimes(1);
    });
    describe('Confirm Button Click Handler', () => {
      beforeEach(() => {
        mockDispatch = jest.fn();
      });
      describe('URL already validated', () => {
        it('dispatches a confirm button clicked action', () => {
          mapDispatchToProps(mockDispatch).confirmClickHandler(true);
          expect(mockDispatch.mock.calls[0][0].type).toEqual(A_CONFIRM_BUTTON_CLICKED);
          expect(mockDispatch).toHaveBeenCalledTimes(1);
        });
      });
      describe('URL not already validated', () => {
        it('dispatches a validate URL action', () => {
          mapDispatchToProps(mockDispatch).confirmClickHandler(false, 'someURL');
          expect(mockDispatch.mock.calls[1][0].type).toEqual(A_VALIDATE_REPO_URL);
          expect(mockDispatch.mock.calls[1][0].data).toEqual({ URL: 'someURL' });
          expect(mockDispatch).toHaveBeenCalledTimes(2);
        });
        it('dispatches a confirm button clicked action', () => {
          mapDispatchToProps(mockDispatch).confirmClickHandler(false);
          expect(mockDispatch.mock.calls[0][0].type).toEqual(A_CONFIRM_BUTTON_CLICKED);
          expect(mockDispatch).toHaveBeenCalledTimes(2);
        });
      });
    });
  });
  describe('componentDidUpdate', () => {
    it('should call informationConfirmed when information has been validated and confirmed', () => {
      const props = {
        inputValidationFailed: () => {},
        URLFieldChange: () => {},
        nameFieldChanged: () => {},
        cancelClickHandler: () => {},
        repoURL: '',
        repoName: '',
        helpMessage: '',
        errorMessage: '',
        informationConfirmed: jest.fn(),
        validated: true,
        confirmed: false,
        isLoading: false,
      };
      const wrapper = shallow(<AddRepoContainer {...props} />);
      wrapper.setProps({ confirmed: true });
      expect(wrapper.instance().props.informationConfirmed).toHaveBeenCalled();
    });
  });
});
