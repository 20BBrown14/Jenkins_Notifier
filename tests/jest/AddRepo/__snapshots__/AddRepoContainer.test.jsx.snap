// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRepoContainer Initialization should render a default view 1`] = `
<AddRepoView
  URLFieldChange={[Function]}
  cancelClickHandler={[Function]}
  confirmClickHandler={[Function]}
  confirmed={false}
  errorMessage=""
  formInvalid={false}
  helpMessage=""
  isLoading={false}
  nameFieldChanged={[Function]}
  repoName=""
  repoURL=""
  validateClickHandler={[Function]}
  validated={false}
/>
`;
